### PR TITLE
Series creation improvements

### PR DESCRIPTION
**Pull request by jwilder**
This fixes a few performance issues I was seeing with series creation.
* There was lock contention on the series file when `LogFile.execSeriesEntry` was run.  We already had the parsed series data further up the stack and `LogFile.execSeriesEntry` would need to look it up again and re-parse it.  I tacked on the parsed series data to the `LogEntry` and `execSeriesEntry` will use it if exists, otherwise fall back to looking it up in the `SeriesFile`.
* Inserting to RHH was creating lots of little allocations.  I switched it to use some temp buffers that can be re-used.
* The `mmap.Map` call `madvise(MADV_RANDOM)` which was trigger a lot of read IOPS on the series files.  I removed that and seems to reduce read IOPS.
* Loading a big TSM file was also trigger lots of read IOPS due to page faults when scanning the index.  I added a hint `madvise(MADV_WILLNEED)` that helps a bit.  I think a better solution would be to store some extra info in the TSM index to avoid scanning at load time.

###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: pkg/rhh/rhh.go [hunk 1]
```diff
@@ -268,6 +270,16 @@ func pow2(v int64) int64 {
 	panic("unreachable")
 }
 
+func assign(x, v []byte) []byte {
+	if cap(x) < len(v) {
+		x = make([]byte, len(v))
+	} else {
```
Comment (e-dard) [inline pkg/rhh/rhh.go]: Could make this function a little shorter like this?

```go
func assign(x, v []byte) []byte {
	if cap(x) < len(v) {
		x = make([]byte, len(v))
	}
	copy(x, v)
	return x[:len(v)]
}
```

File: tsdb/engine/tsm1/reader.go [hunk 2]
```diff
@@ -1374,6 +1375,10 @@ func (m *mmapAccessor) init() (*indirectIndex, error) {
 		return nil, fmt.Errorf("mmapAccessor: invalid indexStart")
 	}
 
+	// Hint to the kernal that we will be reading the file.  It would be better to hint
+	// that we will be reading the index section, but that doesn't seem to work ATM.
+	_ = madvise(m.b, syscall.MADV_WILLNEED)
```
Comment (e-dard) [inline tsdb/engine/tsm1/reader.go]: Could we add this call to the Series File index? We will always need to read from that file too right?

File: pkg/mmap/mmap_unix.go [hunk 3]
```diff
@@ -38,11 +37,6 @@ func Map(path string, sz int64) ([]byte, error) {
 		return nil, err
 	}
 
-	if _, _, err := syscall.Syscall(syscall.SYS_MADVISE, uintptr(unsafe.Pointer(&data[0])), uintptr(len(data)), uintptr(syscall.MADV_RANDOM)); err != 0 {
```
[the remark this answers is not in the document]
Comment (jwilder) [reply pkg/mmap/mmap_unix.go]: I've had to remove any place where we're using `MADV_RANDOM` in the past due to adverse performance problems.  See: https://github.com/influxdata/influxdb/pull/8872 and https://github.com/influxdata/influxdb/pull/5221.  Other projects (https://github.com/boltdb/bolt/issues/691) have seen similar issues.

I only add `madavise` calls in specific spots after testing them quite extensively and can see that they improve what I'm measuring.  Despite what the docs say they do, it's kind of a black box as to what kernel actually does.

File: tsdb/engine/tsm1/reader.go (hunk 2 above)
Comment (jwilder) [reply tsdb/engine/tsm1/reader.go]: I was not measuring series file IOPS so I would not add this there without seeing the actual change in performance.   This does have _some_ benefit with larger TSM files that are not in the page cache though.

File: pkg/mmap/mmap_unix.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (e-dard) [reply pkg/mmap/mmap_unix.go]: @jwilder I guess what I'm saying is: we should do some explicit experimentation in the future on all the sub systems and see if any advice other than `NORMAL` is appropriate anywhere. If we're doing really sequential stuff then `MADV_SEQUENTIAL` might be even better for read IOPS than `MADV_NORMAL` for example.

**Pull request merged**